### PR TITLE
Improve logging and error handling, 

### PR DESCRIPTION
**Pull request by PapaCharlie**
Overhauls the logging and error responses from the `*Conn`. Namely, it changes the logging backend to `log/slog` which makes it easier to manage the library's logging centrally. It also captures the reason why a specific connection to a ZK server was closed in the error that is propagated to all callers. This can be checked with `errors.Is`.

### CODE REVIEW

**Review by bohhyang — APPROVED** [review]

**Review by bohhyang — APPROVED** [review]

**Review by shivamgupta1 — COMMENTED** [review]
Not sure about this one. It's a big change and it's already getting further and further away from the upstream (9 commits behind). It'll be a big pain to merge the upstream changes and we might miss out on important bug fixes and improvements.

**Comment by PapaCharlie** [comment]
Tbqh I don't think the upstream is managed at all, there are no active
maintainers as far as I could tell. I asked LinkedIn to fork this repo so
we would become official maintainers.

**Comment by PapaCharlie** [comment]
Take a look at the repo this repo is itself a fork of:
https://github.com/go-zookeeper/zk , and it itself is behind.

File: conn.go
Comment (shivamgupta1) [inline conn.go]: The  backoff should apply once all the servers in the DNS list have been tried. I remember that's how it used to work in the C client. Do you mind verifying the Java client's behavior?

**Comment by shivamgupta1** [comment]
@PapaCharlie https://github.com/samuel/go-zookeeper is the one that seems unmaintained (last commits were 4 years ago). https://github.com/go-zookeeper/zk seems to have had recent commits (3 months ago). Do you mind double checking this, and confirming that indeed there are no owners? Also, how's the process for becoming official maintainers going to be like?

Nevertheless, it might need to be discussed with the management (if not already done) in case we are trying to take up ownership of a large project that most people in the team don't know much about.

**Comment by PapaCharlie** [comment]
These are the 8 commits that we're behind on:
image
They're all updating the github actions etc, no actual changes. I will merge with the upstream if you want, just to be safe

File: dnshostprovider.go
Comment (shivamgupta1) [inline dnshostprovider.go]: Can a new implementation of the `DNSHostProvider` interface be created rather than modifying existing code? If possible, it should be preferred because of safety and readability.

File: dnshostprovider.go
Comment (shivamgupta1) [inline dnshostprovider.go]: Just a note that this will now throttle retries even if a connection was never established, which is better than the previous implementation, so it looks good. Hope this fact was the motivation here.

File: dnshostprovider_test.go
Comment (bohhyang) [inline dnshostprovider_test.go]: host and addr params should be reversed.

File: dnshostprovider_test.go
Comment (bohhyang) [inline dnshostprovider_test.go]: do we need that many (10k) iterations to hit each of 6 addresses at least once?

File: dnshostprovider.go
Comment (PapaCharlie) [reply dnshostprovider.go]: Right, if there's only one host in the list, it will cause it to back off between retrying that host. That was indeed the intent. However if there's multiple hosts it won't back off in between

File: dnshostprovider.go
Comment (PapaCharlie) [reply dnshostprovider.go]: I did it that way since they were fundamentally not that different but sure, I'll copy it and mark the old one as deprecated. The semantics of the `HostProvider` interface have changed though (`Next` returns a boolean on call earlier), so there will still be some changes.

File: conn.go
Comment (PapaCharlie) [reply conn.go]: Can confirm the Java client only sleeps after trying all the hosts

File: dnshostprovider_test.go
Comment (PapaCharlie) [reply dnshostprovider_test.go]: Ooops, I didn't address this in the other PR. That being said, this test runs in 2ms on my machine. 10k was totally arbitrary but it avoids this becoming somewhat flaky

File: dnshostprovider_test.go
Comment (PapaCharlie) [reply dnshostprovider_test.go]: Nice catch!

File: dnshostprovider.go
Comment (PapaCharlie) [reply dnshostprovider.go]: Done in #6

**Review by bohhyang — APPROVED** [review]

File: dnshostprovider_test.go
Comment (bohhyang) [reply dnshostprovider_test.go]: ACK

**Pull request merged**
